### ISSUE TITLE
InvalidSignatureException when keep-alive connection is retried

### ISSUE THREAD

**Issue by carlhoerberg** [taylorfinnell/awscr-signer#56]
When a long lived HTTP connection times out and the http client has to reconnect the `before_request` will be executed again, but then there are already headers which I think this code will use: https://github.com/taylorfinnell/awscr-signer/blob/master/src/awscr-signer/signers/v4.cr#L83-L85

but that results in a `InvalidSignatureException`. It can be solved by deleting the headers first:

```crystal
      @http.before_request do |request|
        request.headers.delete "Authorization"
        request.headers.delete "X-Amz-Content-Sha256"
        request.headers.delete "X-Amz-Date"
        signer.sign(request)
      end
```

**Comment by jgaskins** [comment]
iiiiinteresting. This sounds similar to the issue I've been having in #52. I'll try your solution in my shard and see if that helps.

**Comment by philipp-classen** [comment]
I ran into the exact problem also when trying to upgrade [easy-awscr](https://github.com/philipp-classen/easy-awscr) (build on top of [awscr-s3](https://github.com/taylorfinnell/awscr-s3)) to make use of HTTP client pooling. I can reliably reproduce on AWS when a connection is being reused.

I'm still trying to get to the core of the problem; but I can confirm that the workaround here to remove the headers seems to work. My intuition is also that it has to be a bug in [awscr-signer](https://github.com/taylorfinnell/awscr-signer). When connection are not reused, but instead new HTTP::Clients are created, it does not happen. But when a connection is reused, it always fails, as far as I can tell. In other words, confirming your analyses so far.

**Comment by philipp-classen** [comment]
I might have jumped to conclusion to fast: looks like the interesting parts happens in the HTTP::Client from the Crystal standard library:
https://github.com/crystal-lang/crystal/blob/fef07780e669bfa79c58e43595aef8c5617ca4e5/src/http/client.cr#L582

```
  private def exec_internal(request)
    implicit_compression = implicit_compression?(request)
    begin
      response = exec_internal_single(request, implicit_compression: implicit_compression)
    rescue exc : IO::Error
      raise exc if @io.nil? # do not retry if client was closed      <------ exc: write (#<TCPSocket:0x7f81f51c1a20>): Broken pipe
      response = nil
    end
    return handle_response(response) if response

    # Server probably closed the connection, so retry once
    close
    request.body.try &.rewind        # <------ restores the body, but not the headers
    response = exec_internal_single(request, implicit_compression: implicit_compression)
    return handle_response(response) if response

    raise IO::EOFError.new("Unexpected end of http response")
  end
```

What happens is that the TCPSocket breaks. (That it *always* breaks in my setup is a different problem; but since sockets *might* break, it is a valid edge case anyway. So let's ignore it here that I may have misconfigured the keepalive part in my setup, or that the S3 backend might not properly handle connection reuse.)

The client now silently closes the connection to try again with a fresh TCPSocket. That makes sense, but the problem is that it only resets the body, but the request headers are still set.

The workaround to clear the headers in the before_request hook takes care of it. Thus, it explains why it works and that it is a practical fix for now. Since I was curious, I tried to modify the `exec_internal` function to first take a snapshot of the headers and then restore it. With that change, the signature errors are also gone.

I will open an issue in the Crystal project to discuss if this behavior could be improved (though my experimental fix in the HTTP::Client and HTTP::Headers involves some extra copying, which may degrade performance). But I would currently see it out of scope of the awscr-signer project.

My summary so far is that it is:
* It is *not* a bug in the awscr-signer logic
* The error recovery in the HTTP::Client should arguably retry with the original state of the headers (to be discussed in the Crystal project). Based on my experience that would solve the problem.
* As a fix in the affected libraries, there are two possible solutions:
  * (recommended IMHO) remove the headers with the before_request hook (as explained already)
  * never reuse connections (e.g. create new HTTP::Client instances or close existing ones if you need to reuse them)

**Comment by philipp-classen** [comment]
Look again into it again (related to the open question whether HTTP headers should be restored, see https://github.com/crystal-lang/crystal/issues/16028). I think, the problem with the current signing logic is when it builds the `canonical_request`, which eventually defines the text that is signed:

First, HTTP request headers are copied over: https://github.com/taylorfinnell/awscr-signer/blob/88f6be8134ca175218f64bbfb0717d84e6cf505e/src/awscr-signer/signers/v4.cr#L98

Then there is a "X-Amz-Content-Sha256" header added (note: not replaced): https://github.com/taylorfinnell/awscr-signer/blob/88f6be8134ca175218f64bbfb0717d84e6cf505e/src/awscr-signer/signers/v4.cr#L105

If HTTP requests start empty, this works. But if they are repeated (and not cleared), then the `X-Amz-Content-Sha256` header, which set in the previous, still exists from the previous time the request was tried. As it has been copied over, "canonical_request.headers.add" now adds it again. In the end, the parameter exists twice. And this will be then used as the text to be signed.

As a test, I changed
```
canonical_request.headers.add("X-Amz-Content-Sha256", canonical_request.digest)
->
canonical_request.headers["X-Amz-Content-Sha256"] ||= canonical_request.digest
```

I doubt that this is a proper fix, but I wanted to understand if signing works then. In my tests, it did indeed fix it.

Probably a cleaner solution would be to tackle it before in the first step whether the headers are copied. My understanding is that it should not copy all headers, especially not the `X-Amz-Content-Sha256` header. Also, I'm not sure if the `canonical_request.headers` key in the canonical_request should ever be set more than once. I'm lacking understanding of the signing protocol, but I would assume that you cannot sign multiple payloads in one request.

Needs more investigation, but I'm leaning now towards treating it (again) as a bug in the library.

---

As a side-note: changing the Crystal HTTP::Client implementation to run the `before_request` hook only once would make the bug also go away (tested with [this diff](https://github.com/crystal-lang/crystal/issues/16028#issuecomment-3135260741)). It is not clear if that would introduce other problems, so it is not a change to be done lightly. In any case, with the current knowledge, I don't think the `awscr-signer` should make any assumptions about it. I'm confident that it can be written in a way that it works correctly when called once or twice and with or without headers getting reset.

**Comment by miry** [comment]
Just a quick clarification to make sure I understand how the HTTP client handles headers.

I assumed that the `HTTP::Headers` structure maps `String -> Array(String)`.  
Given that, should the validation snippet for checking or inserting a header look like this?

```crystal
canonical_request.headers.add("X-Amz-Content-Sha256", canonical_request.digest) unless canonical_request.headers.has_key?("X-Amz-Content-Sha256")
```

@philipp-classen 
do you happen to have an example of how we could simulate this issue in an integration test?  
If so, we could add it under [examples](https://github.com/taylorfinnell/awscr-s3/tree/master/examples) in the `awscr-s3` repo or here.

**Comment by philipp-classen** [comment]
@miry Sorry, I forgot to answer. I think, it is not easy to simulate it in an integration test. At least not the situation that the client wants to reuse a connection, but the server closes it. Maybe it would work to test on a lower level: assuming the signing function is idempotent (not sure though if that property holds), calling it twice should still work. I have not tried, but there is a good chance that it would currently break.

Regarding an example, I think I should be able to provide one. I was also thinking of proposing a patch for the signer, but without a good example that can be run by others, it is difficult. The real-world example that I used so far, is not ideal: first, too complex and somewhat flaky. Second, it builds on private repositories. But I think, it should be possible to reproduce by only uploading some dummy files.

Regarding the question on how it should be fixed: I'm not sure yet. As sketched above, my intuition is that the signing logic should be idempotent. If the input (the message content) does not change, it should result in the same output IMHO. There is a timestamp, but when repeating it should be safe to reuse the old time. The rest should not differ (e.g., checksums).

**Comment by philipp-classen** [comment]
Did not find time and energy to look into it yet. But I see now that Crystal 1.18 got released and changes the behavior now (`before_request` will be called only once per request):

https://github.com/crystal-lang/crystal/pull/16064

That should also eliminate the need for the workarounds. I have not tested against 1.18; but based on my [old comment](https://github.com/crystal-lang/crystal/issues/16028#issuecomment-3138183723), I was no longer able to reproduce in that case.

In summary, there is a pretty good chance that the concrete problem is gone when building with Crystal >=1.18.

Same for the need to have workarounds. Though the existing workarounds should do no harm and are needed for <1.18.
